doajtest/selenium_helpers.py

```diff
@@ -0,0 +1,102 @@
+from multiprocessing import Process
+from typing import TYPE_CHECKING
+
+from selenium import webdriver
+from selenium.webdriver import DesiredCapabilities
+from selenium.webdriver.common.by import By
+
+from doajtest.helpers import DoajTestCase
+from portality import app, models
+
+if TYPE_CHECKING:
+    from selenium.webdriver.remote.webdriver import WebDriver
+
+
+class SeleniumTestCase(DoajTestCase):
+    DOAJ_HOST = '172.17.0.1'
+    DOAJ_PORT = 5014
+    doaj_process = None
+    SELENIUM_URL = 'http://localhost:4444/wd/hub'
+    selenium = None  # selenium driver
+
+    def setUp(self):
+        super().setUp()
+
+        # run doaj server in a background process
+        def _run():
+            app.run_server(host=self.DOAJ_HOST, port=self.DOAJ_PORT)
+
+        self.doaj_process = Process(target=_run)
+        self.doaj_process.start()
+
+        # prepare selenium driver
+        path_chrome_driver = self.app_test.config.get('SELENIUM_CHROME_DRIVER_PATH')
+        if path_chrome_driver:
+            # run selenium with your local browser
+            options = webdriver.ChromeOptions()
+            options.add_argument('--start-maximized')  # maximize browser window
+            browser_driver = webdriver.Chrome(path_chrome_driver, options=options)
+        else:
+            # run selenium with docker remote browser
+            options = webdriver.ChromeOptions()
+            options.add_argument("no-sandbox")
+            options.add_argument("--disable-gpu")
+            options.add_argument("--window-size=800,600")
+            options.add_argument("--disable-dev-shm-usage")
+            browser_driver = webdriver.Remote(
+                command_executor=self.SELENIUM_URL,
+                desired_capabilities=DesiredCapabilities.CHROME,
+                options=options,
+            )
+
+        self.selenium = browser_driver
+        self.selenium.maximize_window()  # avoid something is not clickable
+        self.selenium.implicitly_wait(10)
+
+    def tearDown(self):
+        super().tearDown()
+
+        self.doaj_process.terminate()
+        print('doaj process terminated')
+
+        self.selenium.quit()
+
+    @classmethod
+    def get_doaj_url(cls) -> str:
+        return f'http://{cls.DOAJ_HOST}:{cls.DOAJ_PORT}'
```

Comment:
Likewise, we may want to run this on a service running over `https`, so I'd move the whole `doaj_url` to configuration.

In fact, we have this in `BASE_URL` in settings.py

Reply:
is different with `BASE_URL` 

for every test case (`setUp`) in current selenium structure, 

it run standalone doaj web server in background.
```
# run doaj server in a background process
def _run():
    app.run_server(host=self.DOAJ_HOST, port=self.DOAJ_PORT)
```
therefore, `get_doaj_url` must be `f'http://{cls.DOAJ_HOST}:{cls.DOAJ_PORT}'` instead of configurable 

We might need to run `cls.originals = patch_config(app, {` to patch `BASE_URL` if any test case need `BASE_URL`



-------
about `https`, I would suggest return `f'https://{cls.DOAJ_HOST}:{cls.DOAJ_PORT}'`
```
app.run_server(host=self.DOAJ_HOST, port=self.DOAJ_PORT, fake_https=True)
```

Reply:
ok, all I suggest is that all of that stuff should come out of configuration.  So when you set `DOAJ_HOST` and `DOAJ_PORT` and also whether it uses `http` or `https` needs to come from some settings, so that it can be run on anyone's setup without modifying the source of the tests.

Reply:
But we may run these tests on the test system, which does use https, or in CI which may or may not use https, so I'd prefer we don't hard code the settings from the start

Reply:
by current design, It can't point to other server. because:
* it run doaj web server in background
* all index will be **drop** for each test case
* some file could be already exist / not found in remote server or local store .

is it point to other server is requirement?
if yes I need take sometime to modify the structure.

Reply:
Ah, I see, we can't just run these tests against a running system, they operate like unit tests.

For now we don't need to restructure to do that, we can leave the code as-is.  My main point was purely about having clean code, and hard-coding "http" feels wrong, as we put all other references to urls into configuration.

Reply:
let's hard-coding 'http' for now unit some testcase need to test with https